File: core/src/main/kotlin/org/pastalab/fray/core/concurrency/primitives/SemaphoreContext.kt
package org.pastalab.fray.core.concurrency.primitives

import java.util.concurrent.Semaphore
import org.pastalab.fray.core.ThreadContext
import org.pastalab.fray.core.concurrency.operations.InterruptionType
import org.pastalab.fray.core.concurrency.operations.ThreadResumeOperation
import org.pastalab.fray.rmi.ResourceInfo
import org.pastalab.fray.rmi.ResourceType
import org.pastalab.fray.rmi.ThreadState

class SemaphoreContext(var totalPermits: Int, semaphore: Semaphore) :
    InterruptibleContext,
    Acquirable(ResourceInfo(System.identityHashCode(semaphore), ResourceType.SEMAPHORE)) {
  private val lockWaiters = mutableMapOf<Long, Pair<Int, LockWaiter>>()

  fun acquire(
      permits: Int,
      shouldBlock: Boolean,
      canInterrupt: Boolean,
      thread: ThreadContext
  ): Boolean {
    if (totalPermits >= permits) {
      totalPermits -= permits
      thread.acquiredResources.add(this)
      return true
    } else {
      if (canInterrupt) {
        thread.checkInterrupt()
      }
      if (shouldBlock) {
        lockWaiters[Thread.currentThread().id] = Pair(permits, LockWaiter(canInterrupt, thread))
      }
    }
    return false
  }

  fun drainPermits(): Int {
    val permits = totalPermits
    totalPermits = 0
    return permits
  }

  fun release(permits: Int, threadContext: ThreadContext) {
    totalPermits += permits
    if (totalPermits > 0) {
      threadContext.acquiredResources.add(this)
      val it = lockWaiters.iterator()
      while (it.hasNext()) {
        val (tid, p) = it.next()
        if (totalPermits >= p.first) {
          p.second.thread.pendingOperation = ThreadResumeOperation(true)
          p.second.thread.state = ThreadState.Runnable
          lockWaiters.remove(tid)
        }
      }
    }
  }

  fun reducePermits(permits: Int) {
    totalPermits -= permits
  }

  override fun unblockThread(tid: Long, type: InterruptionType): Boolean {
    val lockWaiter = lockWaiters[tid] ?: return false
    val noTimeout = type != InterruptionType.TIMEOUT
    if (lockWaiter.second.canInterrupt) {
      lockWaiter.second.thread.pendingOperation = ThreadResumeOperation(noTimeout)
      lockWaiter.second.thread.state = ThreadState.Runnable
      lockWaiters.remove(tid)
    }
    return false
  }
}

File: dotnet/src/Fray/Core/Contexts/SemaphoreContext.cs
using Fray.Core.Operations;

namespace Fray.Core.Contexts;

/// <summary>
/// Models a counting semaphore.
///
/// Mirrors <c>org.pastalab.fray.core.concurrency.context.SemaphoreContext</c>.
/// </summary>
public sealed class SemaphoreContext : Acquirable, IInterruptibleContext
{
    private readonly Dictionary<int, (int Permits, LockWaiter Waiter)> _lockWaiters = new();

    public int TotalPermits { get; private set; }

    public SemaphoreContext(int totalPermits, object semaphore)
        : base(new ResourceInfo(ObjectIds.Of(semaphore), ResourceType.Semaphore))
    {
        TotalPermits = totalPermits;
    }

    public bool Acquire(int permits, bool shouldBlock, bool canInterrupt, ThreadContext thread)
    {
        if (TotalPermits >= permits)
        {
            TotalPermits -= permits;
            thread.AcquiredResources.Add(this);
            return true;
        }
        if (canInterrupt)
        {
            thread.CheckInterrupt();
        }
        if (shouldBlock)
        {
            _lockWaiters[thread.Index] = (permits, new LockWaiter(canInterrupt, thread));
        }
        return false;
    }

    public void Release(int permits)
    {
        TotalPermits += permits;
        if (TotalPermits <= 0)
        {
            return;
        }
        foreach (var (index, entry) in _lockWaiters.ToList())
        {
            if (TotalPermits >= entry.Permits)
            {
                entry.Waiter.Thread.PendingOperation = new ThreadResumeOperation(true);
                entry.Waiter.Thread.State = FrayThreadState.Runnable;
                _lockWaiters.Remove(index);
            }
        }
    }

    public int DrainPermits()
    {
        var permits = TotalPermits;
        TotalPermits = 0;
        return permits;
    }

    public bool UnblockThread(ThreadContext thread, InterruptionType type)
    {
        if (!_lockWaiters.TryGetValue(thread.Index, out var entry))
        {
            return false;
        }
        var noTimeout = type != InterruptionType.Timeout;
        if (entry.Waiter.CanInterrupt || type == InterruptionType.Force || type == InterruptionType.Timeout)
        {
            entry.Waiter.Thread.PendingOperation = new ThreadResumeOperation(noTimeout);
            entry.Waiter.Thread.State = FrayThreadState.Runnable;
            _lockWaiters.Remove(thread.Index);
        }
        return false;
    }
}
